app/models/authentication.rb

```diff
@@ -1,11 +1,81 @@
 #encoding: utf-8
 class Authentication < ActiveRecord::Base
+  before_destroy :can_be_destroyed?, prepend: true
+
   FACEBOOK = "facebook"
   GOOGLE = "google_oauth2"
   TWITTER = "twitter"
-  MEETUP="meetup"
-  LINKEDIN="linkedin"
-  OPENID="openid"
-  PARMA="parma"
+  MEETUP = "meetup"
+  LINKEDIN = "linkedin"
+  OPENID = "openid"
+  PARMA = "parma"
+  TECNOLOGIEDEMOCRATICHE = "tecnologiedemocratiche"
   belongs_to :user, class_name: 'User', foreign_key: :user_id
+
+  UNDETACHABLE_PROVIDERS = [ TECNOLOGIEDEMOCRATICHE ]
+
+
+  def self.oauth_raw_info(oauth_data)
+    oauth_data['raw_info'] || # TD
+    oauth_data['extra'].try(:[], 'raw_info') || # Meetup, Twitter, Google, Linkedin, FB
+    oauth_data['info'] # Parma
+  end
+
+  def self.oauth_provider(oauth_data)
+    oauth_data['provider'].to_s
+  end
+
+  def self.oauth_uid(oauth_data)
+    oauth_data['uid'].to_s
+  end
+
+  def self.oauth_user_info(oauth_data)

```

Comment:
this method is very long. can we make it shorter extracting some parts in separate methods?